src/main.rs

```diff
@@ -197,6 +295,29 @@ async fn run(opts: Command) -> anyhow::Result<()> {
 
     let mut n_update_errors = 0;
 
+    loop {
+        select! {
+            ujobres = updates_finished_rx.recv() => {
+                if let Some(UpdateJobResult { worker_id, result }) = ujobres {
+                    if let Err(e) = result {
+                        tracing::error!("Update worker {worker_id} encountered an error: {e}");
+                        n_update_errors += 1;
+                    }
+                } else {
+                    break;
+                }
+            },
+            kjobres = kill_worker_finished_rx.recv() => {
+                if let Some(KillJobResult { result }) = kjobres {
+                    if let Err(e) = result {
+                        tracing::error!("Kill worker encountered an error: {e}");
+                        return Err(anyhow::anyhow!("Test failed, unable to perform lifecycle management"));
+                    }
+                }
+            },
+        }
+    }
+
```

Comment:
is this block async? when does it end?

Reply:
Is async. It ends when one of the given async computations finishes.
So either `updates_finished_rx.recv()` or `kill_worker_finished_rx.recv()`

Link (again): https://tokio.rs/tokio/tutorial/select

Reply:
Note: we loop because `updates_finished_rx.recv()` might return multiple (success) values and we want to read them all.

The `kill_worker_finished_rx.recv()` will only complete before `updates_finished_rx.recv()` if there is an error with server lifetime management.

During normal execution is gets told to stop here (further down):
https://github.com/dice-group/sparql-transactional-test/blob/12dba00bc1c0f44d3acf5de762ce3ebbba32ba85/src/main.rs#L334

Reply:
Are both this loop and the while loop below `while let Some(UpdateJobResult { worker_id, result }) = updates_finished_rx.recv().await`,

Reply:
I was initially confused what you meant, but yeah idk I must have accidentally left the second loop in, will remove